### PR TITLE
Disclosure testing invalid contract type

### PR DESCRIPTION
**Pull request by carlpulley-da**
Test that invalidly typed contracts are rejected
    
CHANGLELOG_BEGIN
    
* Engine/speedy-level tests for explicit disclosure (#14227): test that invalidly typed contracts are rejected.
    
CHANGELOG_END

- [x] Write initial test to prove that unchanged code accepts invalidly typed contracts
- [x] Add in disclosed contract type checking
- [x] Refactor test to demonstrate incorrectly typed contract disclosures are detected/rejected
- [x] Add in additional LF based contract type testing
- [x] rebase PR once https://github.com/digital-asset/daml/pull/14436 is merged
- [x] ensure PR passes in CI

Future PRs will add in additional testing as per https://github.com/digital-asset/daml/issues/14227

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala [hunk 1]
```diff
@@ -181,7 +181,7 @@ private[lf] object Pretty {
             )
           case DisclosurePreprocessing.NonExistentTemplate(templateId) =>
             text(
-              s"Template $templateId does not exist"
+              s"Template $templateId does not exist for the disclosed contract"
             )
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:185]: 1. LookupError pretty return nicer messages
2. 
3. error message about lookup error are more consistent
4. We do not need to extend the Disclosure Error in the future if we rely on some other info in the package

Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:185]: Something I missed in a previous PR.

I would better replace 
```
final case class NonExistentTemplate(templateId: TypeConName) extends Error
```
by 
```
   final case class LookupError(error: language.LookupError) extends Error
```
and replace the lines here by something like:
```
          case LookupError(error) =>
            text(s"${error.pretty} for disclosed contract")
```

Reasons being:
1 - error.pretty has better error messages
2 - we get more consistent error messages about loopuk error message
3 - we will not need to extend  `DisclosurePreprocessing.Error` in the future if the disclosure preprocessing need more info.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:185]: side question, do we check contracts provided with key hash are actual of a type with contract key.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -224,7 +224,7 @@ private[lf] object Speedy {
   @throws[SErrorDamlException]
   private[speedy] def buildDiscTable(
       disclosures: ImmArray[speedy.DisclosedContract],
-      compiledPackages: CompiledPackages,
+      packageInterface: PackageInterface,
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:227]: :+1:

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM.

I have though a comment related to previous PRs.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala (hunk 1 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:185]: > side question, do we check contracts provided with key hash are actual of a type with contract key.

https://github.com/digital-asset/daml/issues/13897 exists to generally investigate validation of disclosed key hashes.

This PR introduces changes that do check that the contract key type, the operation template type and the contract type all match (ref: https://github.com/digital-asset/daml/blob/145a5da6cc723694a5b210a0ba7782f0d552ee11/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala#L1547)

Or did you have some other type of validation in mind that I'm missing here?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 3]
```diff
@@ -2149,6 +2165,27 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
         }
       }
 
+      "a disclosed contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch of a wrongly typed disclosed contract
+        "wrongly typed contract" in {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:2171]: Similar testing for `fetch_by_key` and `lookup_by_key` are not possible as we have no analogous `COERCE_CONTRACT_KEY` command.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -1593,11 +1593,23 @@ private[lf] object SBuiltin {
...
-              val (control, _) = continue(Some(vcoid))
-              control
+              machine.disclosureTable.contractById.get(coid) match {
+                case Some((actualTemplateId, _)) if actualTemplateId == operation.templateId =>
+                  val vcoid = coid.value
+                  continue(Some(vcoid))._1
+
+                case Some((actualTemplateId, _)) =>
+                  throw SErrorDamlException(
+                    IE.WronglyTypedContract(coid.value, operation.templateId, actualTemplateId)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1603]: Throwing an error here definitely makes sense. I do wonder though if this is the right error:

Consider the current cases where we throw WronglyTypedContract: They’re pretty much all user errors. E.g., I called a choice on a contract of a different type. The crucial part is the reason why we get this error is that the participant knows the correct contractid -> contract mapping and that does not match how you tried to use it.

Now here the situation is a bit different. It's not my usage that is wrong, the mapping in the disclosure table is wrong: A key hash for template T maps to a contract of template T'.

So what do you think of throwing something like an `InconsistentDisclosureTable` error or something along those lines to give users more information on what might have gone wrong and how they can fix it?

**Review by cocreature — APPROVED** [review]
Thanks!

**Pull request merged**